### PR TITLE
feat: Add add_link to span api/sdk

### PR DESCRIPTION
**Pull request by dmathieu**
The specification has recently been updated to add the ability to add span links after span creation.

See:

* https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#add-link
* https://github.com/open-telemetry/opentelemetry-specification/pull/3678
* https://github.com/open-telemetry/opentelemetry-specification/pull/3887

This updates the API and SDK to add this new method.

Note: I am not fully sure how you all are handling additive changes to the API. With Ruby being very permissive with method definitions, I assume shipping everything together is fine.
If you think that could cause an issue (like the API being used with a lower version of the SDK), I'm happy to split this PR into two, so the SDK is shipped first, and the API in another release.

### CODE REVIEW

File: api/lib/opentelemetry/trace/span.rb
Comment (kaylareopelle) [inline api/lib/opentelemetry/trace/span.rb:87]: Using a colon matches the documentation style for `Span#add_event`:
https://github.com/open-telemetry/opentelemetry-ruby/blob/6623a3e41534622fb30bd30ec7892ea2e1b4a6b8/api/lib/opentelemetry/trace/span.rb#L81-L85

```suggestion
      # Example:
```

File: sdk/lib/opentelemetry/sdk/trace/span.rb
Comment (kaylareopelle) [inline sdk/lib/opentelemetry/sdk/trace/span.rb:127]: Same suggestion as above:
```suggestion
        # Example:
```

**Review by kaylareopelle — COMMENTED** [review]
Thank you for bringing the Ruby project into compliance with this spec update!

I'll defer to the @open-telemetry/ruby-maintainers to your questions about API and SDK releases. 

I have a few small documentation comments. 

I'm also curious about this line in the spec: 
> Links added after Span creation may not be considered by [Samplers](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#sampler).

Can you tell me how links added using the `add_link` method are ignored by Samplers?

**Comment by dmathieu** [comment]
> Can you tell me how links added using the add_link method are ignored by Samplers?

[SDK samplers](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#sampler), which perform head sampling (not collector samplers, or any other tail sampler) run on span creation. So when `add_link` is called, the head sampling decision has already been made, and is not reconsidered with the new link.

**Comment by kaylareopelle** [comment]
> [SDK samplers](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#sampler), which perform head sampling (not collector samplers, or any other tail sampler) run on span creation. ...

Thank you, @dmathieu. This connects the dots!

**Review by kaylareopelle — APPROVED** [review]

File: sdk/lib/opentelemetry/sdk/trace/span.rb
Comment (fbogsany) [inline sdk/lib/opentelemetry/sdk/trace/span.rb:145]: `@links + [link]` will allocate 2 additional arrays (one to wrap `link` and one for the `+` result). It should be more efficient to `@links << link`, which will append to the `@links` array in place. The potentially inconsistent intermediate state should be invisible to observers due to the mutex.

Comment (dmathieu) [reply sdk/lib/opentelemetry/sdk/trace/span.rb:145]: The links array is frozen in `trim_links`. So in order to do that, I'd have to stop freezing it.

Comment (fbogsany) [reply sdk/lib/opentelemetry/sdk/trace/span.rb:145]: yes - stop freezing it because it is now mutable (it wasn't before). It should be frozen in `finish` instead.

**Review by fbogsany — APPROVED** [review]

**Pull request merged**
